src/App.tsx

```diff
@@ -1,24 +1,198 @@
+  const [loadingId, setLoadingId] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const filterTodos = (initTodos?: Todo[]) => {
+    let filtered;
+
+    if (initTodos) {
+      filtered = [...initTodos];
+    } else {
+      filtered = [...todos];
+    }
+
+    switch (currentFilter) {
+      case TodoFilter.Active:
+        filtered = filtered.filter(todo => !todo.completed);
+        break;
+      case TodoFilter.Completed:
+        filtered = filtered.filter(todo => todo.completed);
+        break;
+      default:
+        break;
+    }
+
+    setVisibleTodos(filtered);
+  };
+
+  const loadTodos = () => {
+    getTodos(USER_ID).then(items => {
+      setTodos([...items]);
+      filterTodos(items);
+    }).catch(() => {
+      setErrors(prevErrors => ({ ...prevErrors, load: true }));
+    });
+  };
+
+  const removeTodo = (todoId: number) => {
+    setLoadingId(prevTodoId => [...prevTodoId, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prevTodos => prevTodos
+          .filter(todo => todo.id !== todoId));
+        setLoadingId([]);
+      })
+      .catch(() => {
+        setErrors(prevErrors => ({
+          ...prevErrors,
+          delete: true,
+        }));
+        setLoadingId([]);
+      });
+  };
+
+  const addTodo = (newTitle: string) => {
+    const maxId = Math.max(...todos.map(todo => todo.id));
+
+    setTempTodo({
+      id: 0,
+      title: newTitle,
+      completed: false,
+      userId: USER_ID,
+    });
+
+    createTodo({
+      id: maxId + 1,
+      title: newTitle,
+      completed: false,
+      userId: USER_ID,
+    })
+      .then((newTodo) => {
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setErrors(prevErrors => ({
+          ...prevErrors,
+          add: true,
+        }));
+        setTempTodo(null);
+      });
+  };
+
+  const changeTodo = (
+    property: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    value: any,
+    todoId: number,
+  ) => {
+    setLoadingId(prevTodoId => [...prevTodoId, todoId]);
+    const oldTodo = todos.find((todo) => todo.id === todoId);
+
+    if (oldTodo) {
+      const updatedTodo: Todo = { ...oldTodo, [property]: value };
+
+      editTodo(updatedTodo, todoId)
+        .then(editedTodo => {
+          setTodos(prevTodos => {
+            const newTodos = [...prevTodos];
+
+            const index = newTodos.findIndex(todo => todo.id === todoId);
+
+            newTodos.splice(index, 1, editedTodo);
+
+            return newTodos;
+          });
+          setLoadingId([]);
+        })
+        .catch(() => {
+          setErrors(prevErrors => ({
+            ...prevErrors,
+            edit: true,
+          }));
+          setLoadingId([]);
+        });
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
+  useEffect(() => {
+    filterTodos();
+  }, [currentFilter, todos]);
```

Comment:
you don't need `visibleTodos` state
general rule if you calculate some value based on other state values/props no need to save it in state

save it in a common variable, wrap in `useMemo` with the same dependency array